servicetalk-transport-netty-internal/src/testFixtures/java/io/servicetalk/transport/netty/internal/EmbeddedDuplexChannel.java

```diff
@@ -0,0 +1,476 @@
+    }
+
+    /**
+     * Awaits completion of {@link #shutdownOutput()}.
+     *
+     * @throws InterruptedException if the current thread was interrupted
+     */
+    public void awaitOutputShutdown() throws InterruptedException {
+        outputShutdownLatch.await();
+    }
+
+    @Override
+    public ChannelFuture shutdownOutput() {
+        return shutdownOutput(newPromise());
+    }
+
+    @Override
+    public ChannelFuture shutdownOutput(final ChannelPromise promise) {
+        if (!isOpen()) {
+            promise.setFailure(new ClosedChannelException());
+            return promise;
+        }
+
+        assert config != null;
+        if (!config.isAllowHalfClosure()) {
+            return close(promise);
+        }
+
+        assert unsafe != null;
+        unsafe.shutdownOutput(promise);
+        return promise;
+    }
+
+    @Override
+    protected void doShutdownOutput() {
+        isOutputShutdown.set(true);
+        outputShutdownLatch.countDown();
+    }
+
+    @Override
+    public boolean isShutdown() {
+        return isInputShutdown() && isOutputShutdown();
+    }
+
+    @Override
+    public ChannelFuture shutdown() {
+        return shutdown(newPromise());
+    }
+
+    @Override
+    public ChannelFuture shutdown(final ChannelPromise promise) {
+        ChannelFuture shutdownOutputFuture = shutdownOutput();
+        if (shutdownOutputFuture.isDone()) {
+            shutdownOutputDone(shutdownOutputFuture, promise);
+        } else {
+            shutdownOutputFuture.addListener((ChannelFutureListener) sof -> shutdownOutputDone(sof, promise));
+        }
+        return promise;
+    }
+
+    private void shutdownOutputDone(final ChannelFuture shutdownOutputFuture, final ChannelPromise promise) {
+        ChannelFuture shutdownInputFuture = shutdownInput();
+        if (shutdownInputFuture.isDone()) {
+            shutdownDone(shutdownOutputFuture, shutdownInputFuture, promise);
+        } else {
+            shutdownInputFuture.addListener((ChannelFutureListener) sif ->
+                    shutdownDone(shutdownOutputFuture, sif, promise));
+        }
+    }
+
+    private static void shutdownDone(ChannelFuture shutdownOutputFuture,
+                                     ChannelFuture shutdownInputFuture,
+                                     ChannelPromise promise) {
+        Throwable shutdownOutputCause = shutdownOutputFuture.cause();
+        Throwable shutdownInputCause = shutdownInputFuture.cause();
+        if (shutdownOutputCause != null) {
+            if (shutdownInputCause != null) {
+                shutdownOutputCause.addSuppressed(shutdownInputCause);
+            }
+            promise.setFailure(shutdownOutputCause);
+        } else if (shutdownInputCause != null) {
+            promise.setFailure(shutdownInputCause);
+        } else {
+            promise.setSuccess();
+        }
+    }
+
+    @Override
+    public Queue<Object> inboundMessages() {
+        if (isInputShutdown.get()) {
```

Comment:
if the input is shutdown real transports will stop producing data (eventually), did you consider force clear/release when shutdown inbound happens to emulate this behavior?

Reply:
It will stop producing new messages, but it should read all pending messages. This is what I emulate using:
```java
pipeline().fireUserEventTriggered(ChannelInputShutdownEvent.INSTANCE);
super.flushInbound();
pipeline().fireUserEventTriggered(ChannelInputShutdownReadComplete.INSTANCE);
```
After that, I make sure that it can not produce more messages discarding any following attempts to write inbound.